### DISCUSSION TITLE
[FEATURE] building storymaps from saved maps

### DISCUSSION THREAD

**Discussion by JinIgarashi** [UNDP-Data/geohub#3515] [Ideas]
### Feature Description

Combining one or more saved maps to generate a story map needs to be implemented in GeoHub.

The below screenshot is current implementation of @undp-data/svelte-maplibre-storymap package.

![image](https://github.com/UNDP-Data/geohub/assets/2639701/eb9e5159-0614-4200-85ab-ed62190a14ab)

A story map consists of following:
- Header information: Title, Subtitle, author, logo image and so on.
- Footer information: Copy right information
- Chapters

Each chapter has information of:
- Markdown content
- Map location info (lng, lat, zoom, bearing, pitch)
- Map animation info
- Map interaction info

User can scroll chapters and interact with maps and contents.

### Basic operational flow

Basic flow of creating a story can be shown as the below figure.

```mermaid
flowchart TD
    A[New story] -->|Click New Story button| B(Select a template)
    B -->C{Select a map}
    C -->|Option 1| D[Select a base map theme]
    C -->|Option 2| E[Select an existing map]
    D -->F(a blank story is created)
    E -->F 
    F -->G(Edit header)
    G -->H(Edit footer)
    H -->I(Add a chapter)
    I -->J{Needs a different map}
    J -->|Yes| K{Select a map}
    K -->|Option 1| L[Select a base map]
    K -->|Option 2| M[Select an existing map]
    J -->|No| N(Edit content)
    L -->N 
    M -->N
    N -->O{need to change layer opacity?}
    O -->|Yes|P[Edit layer opacity for onChapterEnter]
    O -->|No|R
    P -->Q(Edit layer opacity for onChapterExit)
    Q -->R{Add another chapter?}
    R -->|Yes|I
    R -->|No|S(End)
```

### Feature Implementation

Final output of storymap is quite straightforward, however the frontend and backend services to configure storymap is complicated.

Like dataset and map, storymap also needs `Favourite` and `Permission` features. In addition, the following basic features need to be implemented.

- Storymap configuration
  - Add a chapter to a storymap
  - Delete a chapter from a storymap
  - Change an order of chapter in a storymap
  - Set header and footer information for storymap
- Storymap viewer
  - List all storymap (according to selected access level) in a page
  - Show a storymap in a page.

Implementing the below functionalities, the following endpoints and tables are required.

#### Endpoint design

* [x] `/api/storymaps` endpoints
  * GET - get storymaps belong to signed in users.
  * POST - register storymap setting
* [x] `/api/storymaps/{id}` endpoints
  * GET - get storymap settings by ID
  * PUT - update storymap setting
  * DELETE - delete storymap
* [x] `/api/storymaps/{id}/permission`
  * GET - get the list of permission
  * POST - register a permission
  * PUT - update a permission
  * DELETE - delete a permission
* [x] `/api/storymaps/{id}/star`
  * POST - add star to storymap
  * DELETE - delete star from storymap
* [x] `/api/storymaps/{id}/star/count`
  * GET - get the number of stars

#### Frontend

* [x] create a storymap plugin
* [x] create frontend page to show configured storymap (`/storymaps/{id}`)
* [x] create front page to search storymaps (`/storymaps`)
* [ ] create frontend page to configure storymap (`/storymaps/{id}/edit`

- My rough design idea of how we can configure storymap in GUI

https://www.figma.com/design/yYnglAN83X9QWpSzBKSY56/GeoHub?node-id=180-2027&t=LpKGFzw4DNo0yZOC-0

#### Table design
see the options of storymaps here. https://github.com/digidem/maplibre-storymap

- storymap: main table for storymap
  - interface: https://github.com/UNDP-Data/geohub/blob/develop/packages/svelte-maplibre-storymap/src/lib/interfaces/StoryMapConfig.ts
  -  templates are predefined and included in svelte-maplibre-storymap plugin.
- storymap_chapters: manage ordering of chapters in storymap
- storymap_chapter: manage a chapter's setting
  - interface: https://github.com/UNDP-Data/geohub/blob/develop/packages/svelte-maplibre-storymap/src/lib/interfaces/StoryMapChapter.ts
  - User can select a blank map style from available base map styles, or can select an existing map saved in GeoHub.
- storymap_favourite: star functionality
- storymap_permission: permission functionality
- storymap_template: 
  - the table can manage CSS text for storymap template

```mermaid
erDiagram
    storymap {
        uuid id PK
        string title
        bytea logo
        string subtitle
        string byline
        string footer
        integer access_level
        string template_id
        integer style_id FK
        string base_style_id
        timestamp createdat
        string created_user
        timestamp updatedat
        string updated_user
    }

    storymap_chapter {
        uuid id PK
        string title
        string description
        bytea image
        string image_alignment
        string alignment
        boolean mapInteractive
        string mapNavigationPosition
        string mapAnimation
        boolean rotateAnimation
        boolean spinGlobe
        boolean hidden
        geometry center
        integer zoom
        double bearing
        double pitch
        integer style_id FK
        string base_style_id
        jsonb on_chapter_enter
        jsonb on_chapter_exit
        timestamp createdat
        string created_user
        timestamp updatedat
        string updated_user
    }

    storymap_chapters {
        uuid storymap_id PK, FK
        uuid chapter_id PK, FK
        integer sequence PK
    }

    storymap_favourite {
        uuid storymap_id PK, FK
        string user_email PK
        timestamp savedat
    }

    storymap_permission {
        uuid storymap_id PK, FK
        string user_email PK
        smallint permission
        timestamp createdat
        timestamp updatedat
    }

    style {
        integer id PK, FK
    }

    storymap ||--|{ storymap_chapters : is
    storymap_chapters }|--|| storymap_chapter : has
    storymap ||--o{ storymap_favourite : has
    storymap ||--o{ storymap_permission : has
    style ||--|{ storymap_chapter : has
    style ||--|{ storymap : has
```

#### Parameters can be changed in GUI

- Header

| parameter | description |
| --- | --- |
| title |  The title of the overall story  |
| logo  |  binary data of a logo image to the header of your story. (Optional)  |
|  subtitle | A subtitle for the story. (Optional) |
|  byline |  Credit the author of the story. Default is created user name |
| template | storymap CSS template ID. Either 'light' or 'dark' currently. Have to be predefined and included in the plugin |
| style_id |  existing style ID. Either style_id or base_style_id is required. |
| base_style_id | base style ID. it will be applied if style_id is not set. |

- Footer

| parameter | description |
| --- | --- |
| footer |  Citations, credits, etc. that will be displayed at the bottom of the story.  |

- Chapter

| parameter | description |
| --- | --- |
| title |  The title of the section, displayed in an h3 element. |
| description |  The main story content for the section in markdown format. |
| image| binary data for image for the chapter (optional) |
| image_alignment | either 'center' or 'left' or 'right' . center shows it under title. left shows left side of markdown. right shows right side of markdown. |
| alignment | either 'center' or 'left' or 'right' or 'full' |
| hidden |  if true, make the slide hidden |
| mapInteractive |  If true, navigation control is enabled. |
| mapNavigationPosition | the position of navigation control. either 'top-right', 'top-left', 'bottom-right' or 'bottom-left' |
| location | location info of chapter. can store center, zoom, pitch and bearing |
| mapAnimation | add animation to move to next chapter. either 'flyTo', 'easeTo' or 'jumpTo' |
| rotateAnimation | if true, Starts a slow rotation animation at the end of the map transition when set to true. The map will rotate 90 degrees over 24 seconds. |
| spinGlobe | if true, Start spinning globe anitation. The map will rotate 360 degrees over 20 seconds. |
| style_id |  existing style ID. Optional. |
| base_style_id | base style ID. it will be applied if style_id is not set. |
| onChapterEnter | Layers to be displayed/hidden/muted when the section becomes active. Array of layer event objects |
| onChapterExit | Same as onChapterEnter except it is triggered when the section becomes inactive. Array of layer event objects |

For `onChapterEnter` and `onChapterExit`, for each object in an array consists of the following properties:
- layer: Layer name as assigned in MapLibre style.
- opacity: The opacity to display the layer. 0 is fully transparent, 1 is fully opaque.
- duration: The length of the opacity transition, numeric, in milliseconds. Default is 300. This is an optional parameter and can be omitted.

For example, we can set objects like below.

```js
{
      onChapterEnter: [
	      {
		      layer: '2e7e25c7-f5db-43a2-b7ba-70af559b623c', // heatmap
		      opacity: 0,
		      duration: 0
	      },
	      {
		      layer: 'e4032157-10e9-4c4a-a69e-efcc5e50ca08', // circle
		      opacity: 0
	      }
      ],
      onChapterExit: [
	      {
		      layer: '2e7e25c7-f5db-43a2-b7ba-70af559b623c', // heatmap
		      opacity: 0
	      },
	      {
		      layer: 'e4032157-10e9-4c4a-a69e-efcc5e50ca08', // circle
		      opacity: 1
	      }
      ]
}
```

If entering to this chapter, make heatmap and circle invisible. and make circle visible when leaving the chapter.

## references

*  https://github.com/digidem/maplibre-storymap
* https://maplibre.org/maplibre-gl-js/docs/examples/scroll-fly-to/

**Comment by JinIgarashi** [comment]
Figma mock up designed by @balagurova is [here](https://www.figma.com/design/xPemdxTmvPi2homUZ2dyw0/UNDP-GeoHub-UI?node-id=3122-66)

**Comment by iferencik** [comment]
Storymaps can be very effective communication tools  if done correctly. At the same time, the fact there is no UI to build them is a proof how difficult working with storymaps can be. 
In essence a storymap is **one or more interconnected map contexts accompanied by various content like text, images, graphs and links**.

Currently building a storymap is a matter of coding and in general it entails using one or more maplibre/mapbox public styles/maps and custom geospatial layers. In Geohub maps are first class citizens (addresable json documents) and the user could potentially take two very different paths when building a storymapn uising UI.

1. Creating the map documents/style **a priori** using GeoHub and import/use them in the storymap UI

    1. advantages
            - simpler UI and app in general
            - clear structure folowing a separation of concern
            - loose coupling
         
    3.  disadvantages
        - many map documents might be required (global extent)
        - longer overall time to produce a sgtorymap
        - some storymaps coule be invalidated by invalid maps
        
6. Add/style geospatial layers **directly** to the story map

    1. advantages
        - storymaps are self contained and do  not depend on map documents or styles
        - tight coupling between a story map and data layers
        - shorter amount of time will be required to build a storymap
        
    2. disadvantages
       - very complex app and UI
       - difficult to learn
       

There two approaches wil generate completely different user experience and probably result in different applications.

**Reply by JinIgarashi** [reply]
My proposed design is for your first option.

**Reply by balagurova** [reply]
I think it's better to avoid building full map creation into the story map (Option 1). It might be too complicated and could overwhelm users, and it's technically much more difficult. An analogy (maybe not perfect but still) is how PowerPoint doesn't offer complex image editing, so users need to edit/create photos before they want to create a presentation. That said, I think it is important to keep the function so users can select maps from public and private ones and create simple map views without any data.

**Reply by iferencik** [reply]
@JinIgarashi and @balagurova we can agree going with option one where we separate the building of a map and using it inside the storymap builder.  I listed the advantages and disadvantages and we need to consider them along the road, that is, if any requests/features will go against them they will no be implemented.
